### PR TITLE
Runner overhaul!

### LINKED ISSUE #203: Add a data processor abstraction to process raw data and convert it into a list of Records in the runners.
This is to convert one line in the raw data to the format that blurr understands (as the Record class).

The simple case is when the input file has one json dict per line that corresponds to one event. For more complex cases one input line could correspond to multiple events (like in ipfix format). The data processor is an abstraction to handle that in the runner.

### LINKED ISSUE #204: Add unit test support to spark runner.

### PR DESCRIPTION
**Pull request by prashastk**
- Added a Runner ABC class to capture the common code and interface for runners we have so far.
- Added unit tests for SparkRunner.
- Added a DataProcessor abstraction to process raw data into a list of Records.
- Updated CLI to support specifying the runner and the data processor to use.

Fixes #203 
Fixes #204

### CODE REVIEW

**Comment by prashastk** [comment]
W00t! With this PR the test coverage is now 93%+!!

File: blurr/__main__.py [hunk 1]
```diff
@@ -21,10 +21,19 @@
...
     --version                   Show version.
+    --runner=<runner>           The runner to use for the transform.
+                                Possible values:
+                                local - Transforms done in memory. <default>
+                                spark - Transforms done using spark locally.
     --streaming-dtc=<dtc-file>  Streaming DTC file to use.
     --window-dtc=<dtc-file>     Window DTC file to use.
     --source=<raw-json-files>   List of source files separated by comma
     --debug                     Output debug logs.
+    --data-processor=<data-processor>   Data processor to use to process each record. If none provided
```
Comment (vigbk) [inline blurr/__main__.py]: Extra words "if none provided the" ?

File: tests/runner/spark_runner_test.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            '}]') in output_text
+
+
+def test_write_output_file_with_stream_and_window_dtc_provided(tmpdir):
+    spark_runner = SparkRunner(['tests/data/raw.json'], 'tests/data/stream.yml',
+                               'tests/data/window.yml')
+    window_data = spark_runner.execute()
+    out_dir = tmpdir.join('out')
+    spark_runner.write_output_file(str(out_dir), window_data)
+    output_files = out_dir.listdir(lambda x: x.basename.startswith('part'))
```
Comment (vigbk) [inline tests/runner/spark_runner_test.py]: Can make these lines into a function

File: blurr/runner/runner.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    def _validate_dtc_syntax(self) -> None:
+        validate(self._stream_dtc)
+        if self._window_dtc is not None:
+            validate(self._window_dtc)
+
+    def execute_per_user_events(self, user_events: Tuple[str, List[Tuple[datetime, Record]]]
+                                ) -> List[Union[Tuple[Key, Any], Tuple[str, Any]]]:
+        identity = user_events[0]
+        events = user_events[1]
```
Comment (vigbk) [inline blurr/runner/runner.py]: can use 
```
identity, events = user_events
```

File: blurr/__main__.py (hunk 1 above)
Comment (prashastk) [reply blurr/__main__.py]: Removed.

File: tests/runner/spark_runner_test.py (hunk 2 above)
Comment (prashastk) [reply tests/runner/spark_runner_test.py]: Done.

File: blurr/runner/runner.py (hunk 3 above)
Comment (prashastk) [reply blurr/runner/runner.py]: Done.

File: blurr/runner/local_runner.py [hunk 4]
```diff
@@ -5,99 +5,73 @@
...
-        time = self._stream_transformer_schema.get_time(record)
-
-        self._user_events[identity].append((time, record))
-
     def _consume_file(self, file: str) -> None:
         with open(file) as f:
-            for record in f:
-                self._consume_record(Record(json.loads(record)))
+            for data_str in f:
+                for identity, time_record in self.get_per_user_records(data_str):
```
Comment (indradeep) [inline blurr/runner/local_runner.py]: Recommend using `identity` instead of `user` in the names.

File: blurr/runner/local_runner.py [hunk 5]
```diff
@@ -5,99 +5,73 @@
...
     def execute_for_all_users(self) -> None:
-        for identity, events in self._user_events.items():
-            block_data, window_data = execute_dtc(events, identity, self._stream_dtc,
-                                                  self._window_dtc)
-
-            self._block_data.update(block_data)
-            self._window_data[identity] = window_data
-
-    def execute(self) -> None:
+        for user_events in self._users_events.items():
```
Comment (indradeep) [inline blurr/runner/local_runner.py]: We are processing `records` which need not be `events`.  

`for identity_record in self._identity_records.items():` ?

File: blurr/runner/runner.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class Runner(ABC):
+    def __init__(self, json_files: List[str], stream_dtc_file: str, window_dtc_file: Optional[str],
+                 data_processor: DataProcessor):
+        self._raw_files = json_files
+        self._schema_loader = SchemaLoader()
+        self._data_processor = data_processor
+
+        self._stream_dtc = yaml.safe_load(open(stream_dtc_file))
```
Comment (indradeep) [inline blurr/runner/runner.py]: How is the case where no DTCs are specified handled?  You are setting default value for runner and data processor in the CLI, but I don't see how missing DTCs are handled here.

File: blurr/runner/runner.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        self._window_dtc = None if window_dtc_file is None else yaml.safe_load(
+            open(window_dtc_file))
+        self._validate_dtc_syntax()
+
+        self._stream_dtc_name = self._schema_loader.add_schema(self._stream_dtc)
+        self._stream_transformer_schema = self._schema_loader.get_schema_object(
+            self._stream_dtc_name)
+
+    def _validate_dtc_syntax(self) -> None:
```
Comment (indradeep) [inline blurr/runner/runner.py]: I see this block only being invoked once from the method above.  Why separate it out into a function?

File: blurr/runner/local_runner.py (hunk 5 above)
Comment (prashastk) [reply blurr/runner/local_runner.py]: Done, also changed other instances of 'events' here where are operating on records.

File: blurr/runner/runner.py (hunk 6 above)
Comment (prashastk) [reply blurr/runner/runner.py]: docopts ensures that the streaming DTC is provided. Also if someone calls this directly then its expected to crash right now.

File: blurr/runner/runner.py (hunk 7 above)
Comment (prashastk) [reply blurr/runner/runner.py]: Removed and moved up.

**Comment by prashastk** [comment]
I am not sure why coverage is not being reported here.  Looks good based on https://coveralls.io/jobs/35742654.

Merging the PR.

**Pull request merged**